File: py-phylo2vec/tests/test_io.py
"""Tests for I/O functions."""

import numpy as np
import pytest

from ete4 import Tree

from phylo2vec.base.newick import from_newick, to_newick
from phylo2vec.io.reader import load, load_newick
from phylo2vec.io.writer import save, save_newick
from phylo2vec.utils.matrix import sample_matrix
from phylo2vec.utils.newick import create_label_mapping
from phylo2vec.utils.vector import sample_vector
from .config import MAX_N_LEAVES, MIN_N_LEAVES


def _test_save_and_load(tmp_path, n_leaves, sample_fn):
    """Test the read and write functions for Phylo2Vec objects

    Parameters
    ----------
    tmp_path : Path
        Temporary path for saving files
    n_leaves : int
        Number of leaves
    sample_fn : function
        Function to sample a vector or matrix
    """
    vector_or_matrix = sample_fn(n_leaves)

    csv_path = tmp_path / "test.csv"
    save(vector_or_matrix, csv_path)
    loaded_vector_or_matrix = load(csv_path)
    assert np.array_equal(vector_or_matrix, loaded_vector_or_matrix)

    # Unallowed file extension should trigger an AssertionError
    random_path = tmp_path / "test.random"
    np.savetxt(random_path, vector_or_matrix, delimiter=",")
    with pytest.raises(ValueError):
        save(vector_or_matrix, random_path)
        _ = load(random_path)


@pytest.mark.parametrize("n_leaves", [MIN_N_LEAVES, MAX_N_LEAVES])
def test_save_and_load_vector(tmp_path, n_leaves):
    """Test the read and write functions for Phylo2Vec objects

    Parameters
    ----------
    n_leaves : int
        Number of leaves
    """
    _test_save_and_load(tmp_path, n_leaves, sample_vector)


@pytest.mark.parametrize("n_leaves", [MIN_N_LEAVES, MAX_N_LEAVES])
def test_save_and_load_matrix(tmp_path, n_leaves):
    """Test the read and write functions for Phylo2Vec objects

    Parameters
    ----------
    n_leaves : int
        Number of leaves
    """
    _test_save_and_load(tmp_path, n_leaves, sample_matrix)


def _test_and_save_load_newick(tmp_path, n_leaves, sample_fn):
    vector_or_matrix = sample_fn(n_leaves)

    newick_path = tmp_path / "test.newick"
    save_newick(vector_or_matrix, newick_path)
    m2 = load_newick(newick_path)
    assert np.array_equal(vector_or_matrix, m2)

    # Unallowed file extension should trigger an AssertionError
    random_path = tmp_path / "test.random"
    np.savetxt(random_path, vector_or_matrix, delimiter=",")
    with pytest.raises(ValueError):
        save_newick(vector_or_matrix, random_path)
        _ = load_newick(random_path)


@pytest.mark.parametrize("n_leaves", [MIN_N_LEAVES, MAX_N_LEAVES])
def test_save_and_load_newick_vector(tmp_path, n_leaves):
    """Test the read and write functions for Phylo2Vec objects

    Parameters
    ----------
    n_leaves : int
        Number of leaves
    """
    _test_and_save_load_newick(tmp_path, n_leaves, sample_vector)


@pytest.mark.parametrize("n_leaves", [MIN_N_LEAVES, MAX_N_LEAVES])
def test_save_and_load_newick_matrix(tmp_path, n_leaves):
    """Test the read and write functions for Phylo2Vec objects

    Parameters
    ----------
    n_leaves : int
        Number of leaves
    """
    _test_and_save_load_newick(tmp_path, n_leaves, sample_matrix)


@pytest.mark.parametrize("n_leaves", [MIN_N_LEAVES, MAX_N_LEAVES])
def test_save_newick_with_labels(tmp_path, n_leaves):
    """Test saving a Newick file with labels."""

    tr = Tree()
    tr.populate(n_leaves)
    nw_str = tr.write(parser=9)

    # Create an str-to-str mapping and create an integer Newick
    # key: node, value: label
    nw_int, label_mapping = create_label_mapping(nw_str)

    v = from_newick(nw_int)

    newick_path = tmp_path / "test.newick"

    save_newick(v, newick_path, label_mapping)

    with open(newick_path, "r", encoding="utf-8") as f:
        nw_str2 = f.read().strip()

    assert nw_str == nw_str2


@pytest.mark.parametrize("n_leaves", [MIN_N_LEAVES, MAX_N_LEAVES])
def test_load_newick_vector_from_str(n_leaves):
    v = sample_vector(n_leaves)
    nw_str = to_newick(v)
    v2 = load_newick(nw_str)
    assert np.array_equal(v, v2)


@pytest.mark.parametrize("n_leaves", [MIN_N_LEAVES, MAX_N_LEAVES])
def test_load_newick_matrix_from_str(n_leaves):
    m = sample_matrix(n_leaves)
    nw_str = to_newick(m)
    m2 = load_newick(nw_str)
    assert np.array_equal(m, m2)


class TestSaveEdgeCases:
    def test_save_empty(self):
        v = np.array(0)
        with pytest.raises(ValueError):
            save(v, "test.csv")

    def test_save_ndim3(self):
        # array with 3 dimensions
        t = np.zeros((MIN_N_LEAVES, 3, 1))
        # Check that we raise a ValueError
        with pytest.raises(ValueError):
            save(t, "test.csv")

File: r-phylo2vec/tests/testthat/test_io.R
source("config.R")

test_save_and_load <- function(topology_only) {
  tmp_path <- tempdir()
  for (n_leaves in c(2, MIN_N_LEAVES, MAX_N_LEAVES)) {
    vector_or_matrix <- sample_tree(
      n_leaves,
      ordered = FALSE,
      topology_only = topology_only
    )

    # Save and load via CSV
    # Check that the loaded data matches the original
    csv_path <- file.path(tmp_path, "test.csv")
    save_p2v(vector_or_matrix, csv_path)
    vector_or_matrix2 <- load_p2v(csv_path)
    expect_equal(vector_or_matrix, vector_or_matrix2)

    # Unallowed file extension should trigger an error
    random_path <- file.path(tmp_path, "test.random")
    expect_error(save_p2v(vector_or_matrix, random_path))
  }
}

test_that(desc = "save_and_load_vector", {
  test_save_and_load(topology_only = TRUE)
})

test_that(desc = "save_and_load_matrix", {
  test_save_and_load(topology_only = FALSE)
})

test_save_and_load_newick <- function(topology_only) {
  tmp_path <- tempdir()
  for (n_leaves in c(2, MIN_N_LEAVES, MAX_N_LEAVES)) {
    vector_or_matrix <- sample_tree(
      n_leaves,
      ordered = FALSE,
      topology_only = topology_only
    )

    # Save and load via Newick
    # Check that the loaded data matches the original
    newick_path <- file.path(tmp_path, "test.newick")
    save_newick(vector_or_matrix, newick_path)
    vector_or_matrix2 <- load_newick(newick_path)
    expect_equal(vector_or_matrix, vector_or_matrix2)

    # Unallowed file extension should trigger an error
    random_path <- file.path(tmp_path, "test.random")
    expect_error(save_newick(vector_or_matrix, random_path))
  }
}

test_that(desc = "save_and_load_newick_vector", {
  test_save_and_load_newick(topology_only = TRUE)
})

test_that(desc = "save_and_load_newick_matrix", {
  test_save_and_load_newick(topology_only = FALSE)
})

test_save_newick_with_labels <- function(...) {
  tmp_path <- tempdir()
  for (n_leaves in c(2, MIN_N_LEAVES, MAX_N_LEAVES)) {
    tr <- ape::rtree(n_leaves, rooted = TRUE, equiprob = TRUE, ...)

    nw_str <- ape::write.tree(tr)

    out <- create_label_mapping(nw_str)

    vector_or_matrix <- from_newick(out$newick)

    newick_path <- file.path(tmp_path, "test.newick")
    save_newick(vector_or_matrix, newick_path, labels = out$mapping)

    nw_str2 <- readLines(newick_path, warn = FALSE)

    expect_equal(nw_str, nw_str2)
  }
}

test_that(desc = "save_newick_with_labels_vector", {
  test_save_newick_with_labels(br = NULL)
})

test_that(desc = "save_newick_with_labels_matrix", {
  test_save_newick_with_labels()
})

test_load_newick_from_str <- function(topology_only) {
  for (n_leaves in c(2, MIN_N_LEAVES, MAX_N_LEAVES)) {
    vector_or_matrix <- sample_tree(
      n_leaves,
      ordered = FALSE,
      topology_only = topology_only
    )
    nw_str <- to_newick(vector_or_matrix)
    vector_or_matrix2 <- load_newick(nw_str)
    expect_equal(vector_or_matrix, vector_or_matrix2)
  }
}

test_that(desc = "load_newick_from_str_vector", {
  test_load_newick_from_str(topology_only = TRUE)
})

test_that(desc = "load_newick_from_str_matrix", {
  test_load_newick_from_str(topology_only = FALSE)
})

test_that(desc = "load_nbyNot3", {
  tmp_path <- tempdir()
  csv_path <- file.path(tmp_path, "test.csv")

  # N x 2
  writeLines("0,0.1\n1,0.2", csv_path)
  expect_error(load_p2v(csv_path))

  # N x 4
  writeLines("0,0.1,0,0.5\n1,0.2,1,0.6", csv_path)
  expect_error(load_p2v(csv_path))
})


test_that(desc = "save_empty_vector", {
  tmp_path <- tempdir()
  csv_path <- file.path(tmp_path, "test.csv")

  # Triggers the stop arm of save_p2v
  expect_error(save_p2v(c(), csv_path))
})

test_that(desc = "save_empty_matrix", {
  tmp_path <- tempdir()
  csv_path <- file.path(tmp_path, "test.csv")

  # Triggers the stop arm of save_p2v
  expect_error(save_p2v(matrix(array()), csv_path))

  # NOTE: This triggers a different error: panic error in check_m
  # expect_error(save_p2v(matrix(array(0)), csv_path))
})

test_that(desc = "save_ndim3", {
  tmp_path <- tempdir()
  csv_path <- file.path(tmp_path, "test.csv")

  x <- array(runif(60), dim = c(3, 4, 5))

  expect_error(save_p2v(x, csv_path))
})
